src/aws_signature_builder/mod.rs

```diff
@@ -0,0 +1,390 @@
+    let datestamp = now.format("%Y%m%d").to_string();
+    return (amzdate, datestamp)
+}
+
+pub fn generate_aws_signature_headers(
+    query: String,
+    headers: HashMap<String, String>,
+    port: Option<u16>,
+    host: String,
+    amzdate: String,
+    method: String,
+    data: Vec<u8>,
+    security_token: Option<String>,
+    data_binary: bool,
+    datestamp: String,
+    service: String,
+    region: String,
+    access_key: String,
+    secret_key: String,
+    canonical_uri: String) -> HashMap<String, String> {
+    let (canonical_request,
+         payload_hash,
+         signed_headers) = task_1_create_a_canonical_request(
+         query, headers, port, host, amzdate.clone(), method, data, security_token.clone(),
+         data_binary, canonical_uri);
+    let (string_to_sign,
+         algorithm,
+         credential_scope) = task_2_create_the_string_to_sign(
+         amzdate.clone(), datestamp.clone(), canonical_request, service.clone(), region.clone());
+    let signature = task_3_calculate_the_signature(
+        datestamp, string_to_sign, service, region, secret_key);
+    let new_headers = task_4_build_auth_headers_for_the_request(
+        amzdate, payload_hash, algorithm, credential_scope, signed_headers, signature, access_key,
+        security_token);
+    return new_headers;
+}
+
+/// ************* TASK 1: CREATE A CANONICAL REQUEST *************
+/// http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+///
+/// Step 1 is to define the verb (GET, POST, etc.)--already done.
+///
+/// Step 2: Create canonical URI--the part of the URI from domain to query string (use '/' if no
+/// path)
+fn task_1_create_a_canonical_request(
+    query: String,
+    headers: HashMap<String, String>,
+    port: Option<u16>,
+    host: String,
+    amzdate: String,
+    method: String,
+    data: Vec<u8>,
+    security_token: Option<String>,
+    data_binary: bool,
+    canonical_uri: String) -> (String, String, String) {
+
+    // Step 3: Create the canonical query string. In this example (a GET request), request
+    // parameters are in the query string. Query string values must be URL-encoded (space=%20). The
+    // parameters must be sorted by name.  For this example, the query string is pre-formatted in
+    // the request_parameters variable.
+    let canonical_querystring = normalize_query_string(query);
+
+    // If the host was specified in the HTTP header, ensure that the canonical headers are set
+    // accordingly
+    let fullhost = if headers.contains_key("host") {
+        headers["host"].clone()
+    } else {
+        let fullhost = match port {
+            Some(p) => format!("{}:{}", host, p.to_string()),
+            None => host,
+        };
+        fullhost
+    };
+
+    // Step 4: Create the canonical headers and signed headers. Header names and value must be
+    // trimmed and lowercase, and sorted in ASCII order.  Note that there is a trailing \n.
```

Comment:
Actually, this makes it even more obvious that it was directly copied from https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html.  I bet only Amazon knows the answer to that...

I should probably put a note in the README and link to that page from many of these comments to make it clear this is the situation.

Reply:
I'm not shocked that the implementation of a proprietary protocol like this was copy/pasted from a docs page rather than implemented from an actual spec.  It makes me question how feasible it would be to actually do this by following the Amazon docs.